overlord/hookstate/ctlcmd/ctlcmd.go

```diff
@@ -0,0 +1,109 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+// Package ctlcmd contains the various snapctl subcommands.
+package ctlcmd
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/overlord/hookstate"
+
+	"github.com/jessevdk/go-flags"
+)
+
+type baseCommand struct {
+	stdout  io.Writer
+	stderr  io.Writer
+	context *hookstate.Context
+}
+
+func (c *baseCommand) setStdout(w io.Writer) {
+	c.stdout = w
+}
+
+func (c *baseCommand) printf(format string, a ...interface{}) {
+	if c.stdout != nil {
+		c.stdout.Write([]byte(fmt.Sprintf(format, a...)))
+	}
+}
+
+func (c *baseCommand) setStderr(w io.Writer) {
+	c.stderr = w
+}
+
+func (c *baseCommand) errorf(format string, a ...interface{}) {
+	if c.stderr != nil {
+		c.stderr.Write([]byte(fmt.Sprintf(format, a...)))
+	}
+}
+
+func (c *baseCommand) setContext(context *hookstate.Context) {
+	c.context = context
+}
+
+func (c *baseCommand) getContext() *hookstate.Context {
+	return c.context
+}
+
+type ctlCommand interface {
+	setStdout(w io.Writer)
+	setStderr(w io.Writer)
+	setContext(context *hookstate.Context)
+	getContext() *hookstate.Context
+
+	Execute(args []string) error
+}
+
+type commandGenerator func() ctlCommand
+
+var commandGenerators map[string]commandGenerator
```

Comment:
Compound suggestion:

`var commandGenerators = make(map[string]func() command)`

 1. `ctlCommand` => `command` and use `cmd` as the variable name
 2. drop `commandGenerator` and use `func() command`, shorter and more clear
 3. inline the map creation at declaration time, so no need to test for `nil` below